### PR TITLE
refactor: Support EIP-712 for nested structs

### PR DESCRIPTION
**Pull request by ziemen4**
- Refactor to support EIP-712 for nested types

## Summary by CodeRabbit

- **New Features**
	- Introduced a new input structure for handling token transfers, enhancing the contract's data integrity.
	- Updated functions to support the new input format for improved permit processing.

- **Bug Fixes**
	- Adjusted message hashing to accurately reflect the new input structure, ensuring reliable EIP-712 signing.

### CODE REVIEW

File: contracts/src/UntronIntents.sol [hunk 1]
```diff
@@ -42,9 +42,12 @@ abstract contract UntronIntents is IUntronIntents, Initializable {
     bytes32 internal constant EIP712_DOMAIN_TYPEHASH =
         keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)");
 
+    /// @dev EIP-712 input type hash
+    bytes32 internal constant INPUT_TYPEHASH = keccak256("Input(address token,uint256 amount)");
```
[the remark this answers is not in the document]
Comment (ziemen4) [reply contracts/src/UntronIntents.sol:46]: This is included in the inherited interface.

**Comment by ziemen4** [comment]
@coderabbitai resolve

**Comment by alexhooketh** [comment]
@ziemen4 could you take a look at whether what bot suggests (add Input into Intent typehash) makes sense as per EIP-712? I think I've seen something like this in Permit2

**Comment by ziemen4** [comment]
> @ziemen4 could you take a look at whether what bot suggests (add Input into Intent typehash) makes sense as per EIP-712? I think I've seen something like this in Permit2

Yes input is already part of intent typehash. Bot just wants to make it more readable

**Review by alexhooketh — APPROVED** [review]
lgtm

**Pull request merged**